skll/data/readers.py

```diff
@@ -238,12 +241,127 @@ def feat_dict_generator():
                 for ex_num, (_, _, feat_dict) in enumerate(self._sub_read(f)):
                     yield feat_dict
                     if ex_num % 100 == 0:
-                        self._print_progress('{:.8}%'.format(100 * ((ex_num /
-                                                                     total))))
+                        self._print_progress('{:.8}%'.format(100 * ((ex_num / total))))
                 self._print_progress("100%")
 
+        # extract the features dictionary
+        features = feat_dict_generator()
+
+        return ids, labels, features
+
+    def _parse_dataframe(self, df, id_col, label_col, features=None):
+        """
+        Parse the data frame into ids,
+        labels, and features.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            The pandas data frame to parse.
+        id_col : str or None
+            The id column.
+        label_col : str or None
+            The label column.
+        features : list of dict or None
+            The features, if they already exist;
+            if not, then they will be extracted
+            from the data frame.
+            Defaults to None.
+
+        Returns
+        -------
+        ids : np.array
+            The ids for the feature set.
+        labels : np.array
+            The labels for the feature set.
+        features : list of dicts
+            The features for the feature set.
+        """
+        if df.empty:
+            raise ValueError("No features found in possibly "
+                             "empty file '{}'.".format(self.path_or_list))
+
+        # if the id column exists,
+        # get them from the data frame and
+        # delete the column
+        if id_col is not None:
+            ids = df[id_col]
```

Comment:
Good catch. What I _should_ be doing here is checking `if id_col is not None and id_col in df`, and then forcing the `id_col` to `None` otherwise. I'll update.